k9mail-library/src/main/java/com/fsck/k9/mail/transport/SmtpTransport.java

```diff
@@ -756,6 +829,71 @@ private void saslAuthCramMD5(String username, String password) throws MessagingE
         }
     }
 
+    private void saslXoauth2(String username) throws MessagingException, IOException {
+        retryXoauthWithNewToken = true;
+        try {
+            attemptXoauth2(username);
+        } catch (NegativeSmtpReplyException negativeResponse) {
+            if (negativeResponse.getReplyCode() != SMTP_AUTHENTICATION_FAILURE_ERROR_CODE) {
+                throw negativeResponse;
+            }
+
+            oauthTokenProvider.invalidateToken(username);
+
+            if (!retryXoauthWithNewToken) {
+                handlePermanentFailure(negativeResponse);
+            } else {
+                handleTemporaryFailure(username, negativeResponse);
+            }
+        }
+    }
+
+    private void handlePermanentFailure(NegativeSmtpReplyException negativeResponse) throws AuthenticationFailedException {
+        throw new AuthenticationFailedException(negativeResponse.getMessage(), negativeResponse);
+    }
+
+    private void handleTemporaryFailure(String username, NegativeSmtpReplyException negativeResponseFromOldToken)
+        throws IOException, MessagingException {
+        // Token was invalid
+
+        //We could avoid this double check if we had a reasonable chance of knowing
+        //if a token was invalid before use (e.g. due to expiry). But we don't
+        //This is the intended behaviour per AccountManager
+
+        Log.v(LOG_TAG, "Authentication exception, re-trying with new token", negativeResponseFromOldToken);
+        try {
+            attemptXoauth2(username);
+        } catch (NegativeSmtpReplyException negativeResponseFromNewToken) {
+            if (negativeResponseFromNewToken.getReplyCode() != SMTP_AUTHENTICATION_FAILURE_ERROR_CODE) {
+                throw negativeResponseFromNewToken;
+            }
+
+            //Okay, we failed on a new token.
+            //Invalidate the token anyway but assume it's permanent.
+            Log.v(LOG_TAG, "Authentication exception for new token, permanent error assumed",
+                    negativeResponseFromNewToken);
+
+            oauthTokenProvider.invalidateToken(username);
+
+            handlePermanentFailure(negativeResponseFromNewToken);
```

Comment:
this whole block duplicates a lot of of the previous method. maybe encapsulate the general approach, and see if it could also be used in `ImapConnection`.